### ISSUE TITLE
[Feature] Implement LazyLinear / Dense layer that can inference input dim of a linear layer

### ISSUE THREAD

**Issue by PRESIDENT810** [ml-explore/mlx#437]
Both PyTorch and JAX supports this lazy linear layer that can automatically infer what will be the input_dim to initialize using the input tensor's last dimension, and user only needs to specify the output_dim (features). However, in mlx it looks like we currently do not have this feature, and we still need to explicitly specify the input_dim in mlx.nn.Linear

Related APIs
https://flax.readthedocs.io/en/latest/api_reference/flax.linen/_autosummary/flax.linen.Dense.html
https://pytorch.org/docs/stable/generated/torch.nn.LazyLinear.html#torch.nn.LazyLinear

**Comment by awni** [comment]
I'm not opposed to a lazy linear layer, would be pretty reasonable to implement this as a derived class from linear.

**Comment by nakranivaibhav** [comment]
hey @awni  is the feature still requested. I can work on it. What can i call the newclas LazyLinear??

**Comment by nakranivaibhav** [comment]
@awni instead of a seperate class can we have a check in the linear class itself that the input dims have been provided or not  and the code can act accordingly. Just a thought

**Comment by awni** [comment]
Closing this as we decided not to move forward with the implementation.
